### PR TITLE
Add Grow Cut tool

### PR DESCRIPTION
**Pull request by allywarner**
Please test new tool for timing and correctness!

### CODE REVIEW

**Comment by dcwhite** [comment]
@allywarner Lots of merge conflicts to fix...

**Comment by allywarner** [comment]
> @allywarner Lots of merge conflicts to fix...

@dcwhite I've merged them in!

**Comment by allywarner** [comment]
Closes #370

File: src/Application/Tools/Algorithm/FastGrowCut.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  std::vector<LabPixelType> m_imLab;
+  std::vector<FPixelType> m_imDist;
+
+  std::vector<long> m_imSize;
+  long m_DIMX, m_DIMY, m_DIMZ, m_DIMXY, m_DIMXYZ;
+  std::vector<int> m_indOff;
+  std::vector<unsigned char>  m_NBSIZE;
+
+  FibHeap* m_heap;
+  HeapNode* m_hpNodes;
```
Comment (dcwhite) [inline src/Application/Tools/Algorithm/FastGrowCut.h:82]: I know this is ancient code (1996) that you just copied from somewhere, but the memory management of these data structures is unacceptable for 2021 (many raw new/deletes in the client code). I'll checkout the branch and see if I can clean these up.

**Review by dcwhite — CHANGES_REQUESTED** [review]

File: src/Application/Tools/Algorithm/FastGrowCut.h (hunk 1 above)
Comment (allywarner) [reply src/Application/Tools/Algorithm/FastGrowCut.h:82]: Thank you @dcwhite!

Comment (dcwhite) [reply src/Application/Tools/Algorithm/FastGrowCut.h:82]: Please test this branch, and if it works OK, just merge it into the PR branch
https://github.com/SCIInstitute/Seg3D/tree/growcut-improved-memory

Comment (allywarner) [reply src/Application/Tools/Algorithm/FastGrowCut.h:82]: Work okay as far as I can tell!

File: src/Application/Tools/Algorithm/fibheap.cc [hunk 2]
```diff
@@ -0,0 +1,656 @@
+#define _FIBHEAP_CPP
+
+//***************************************************************************
+// This Fibonacci heap implementation is Copyright (c) 1996 by John Boyer.
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/fibheap.cc:4]: Is this lifted strait or is it modified?  we should include a link to the repo: https://github.com/ljzhu/FastGrowCut.  If you changed it enough, we should change the copyright.

File: src/Application/Tools/Algorithm/fibheap.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
+#ifndef FIBHEAP_H
+#define FIBHEAP_H
+
+//***************************************************************************
+// The Fibonacci heap implementation contained in FIBHEAP.H and FIBHEAP.CPP
+// is Copyright (c) 1996 by John Boyer
+//
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/fibheap.h:7]: same as previous

File: src/Application/Tools/Algorithm/itkFastGrowCut.h [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+ */
+
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/itkFastGrowCut.h:28]: // adapted from https://github.com/ljzhu/FastGrowCut

File: src/Application/Tools/Algorithm/itkFastGrowCut.hxx [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+ */
+
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/itkFastGrowCut.hxx:28]: // adapted from https://github.com/ljzhu/FastGrowCut

File: src/Application/Tools/Algorithm/utilities.h [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+ */
+
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/utilities.h:28]: // adapted from https://github.com/ljzhu/FastGrowCut

File: src/Application/Tools/Algorithm/FastGrowCut.hxx [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+ */
+
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/FastGrowCut.hxx:29]: // adapted from https://github.com/ljzhu/FastGrowCut

File: src/Application/Tools/Algorithm/FastGrowCut.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+   THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+   DEALINGS IN THE SOFTWARE.
+ */
+
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/FastGrowCut.h:28]: // adapted from https://github.com/ljzhu/FastGrowCut

File: src/Application/Tools/Algorithm/HeapNode.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
+//
+//  HeapNode.h
+//  
+//
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/HeapNode.h:4]: add copyright info and keep author

File: src/Application/Tools/Algorithm/HeapNode.cc [hunk 10]
```diff
@@ -0,0 +1,39 @@
+//
+//  HeapNode.cpp
+//  
+//
```
Comment (jessdtate) [inline src/Application/Tools/Algorithm/HeapNode.cc:4]: add copyright and keep author

**Review by jessdtate — DISMISSED** [review]

File: src/Application/Tools/Algorithm/fibheap.cc (hunk 2 above)
Comment (allywarner) [reply src/Application/Tools/Algorithm/fibheap.cc:4]: I didn't change this code.

File: src/Application/Tools/Algorithm/fibheap.h (hunk 3 above)
Comment (allywarner) [reply src/Application/Tools/Algorithm/fibheap.h:7]: I didn't change this code.

File: src/Application/Tools/Algorithm/fibheap.cc (hunk 2 above)
Comment (allywarner) [reply src/Application/Tools/Algorithm/fibheap.cc:4]: Should I add our copyright?

Comment (jessdtate) [reply src/Application/Tools/Algorithm/fibheap.cc:4]: no, just a link

**Review by jessdtate — APPROVED** [review]

**Review by dcwhite — APPROVED** [review]

**Pull request merged**